cpp/src/io/parquet/rle_stream.cuh

```diff
@@ -0,0 +1,357 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "parquet_gpu.hpp"
+#include <cudf/detail/utilities/integer_utils.hpp>
+
+namespace cudf {
+namespace io {
+namespace parquet {
+namespace gpu {
+
+// TODO: consider if these should be template parameters to rle_stream
+constexpr int num_rle_stream_decode_threads = 512;
+// the -1 here is for the look-ahead warp that fills in the list of runs to be decoded
+// in an overlapped manner. so if we had 16 total warps:
+// - warp 0 would be filling in batches of runs to be processed
+// - warps 1-15 would be decoding the previous batch of runs generated
+constexpr int num_rle_stream_decode_warps = (num_rle_stream_decode_threads / 32) - 1;
+constexpr int run_buffer_size             = (num_rle_stream_decode_warps * 2);
+constexpr int rolling_run_index(int index) { return index % run_buffer_size; }
+
+/**
+ * @brief Read a 32-bit varint integer
+ *
+ * @param[in,out] cur The current data position, updated after the read
+ * @param[in] end The end data position
+ *
+ * @return The 32-bit value read
+ */
+inline __device__ uint32_t get_vlq32(const uint8_t*& cur, const uint8_t* end)
+{
+  uint32_t v = *cur++;
+  if (v >= 0x80 && cur < end) {
+    v = (v & 0x7f) | ((*cur++) << 7);
+    if (v >= (0x80 << 7) && cur < end) {
+      v = (v & ((0x7f << 7) | 0x7f)) | ((*cur++) << 14);
+      if (v >= (0x80 << 14) && cur < end) {
+        v = (v & ((0x7f << 14) | (0x7f << 7) | 0x7f)) | ((*cur++) << 21);
+        if (v >= (0x80 << 21) && cur < end) {
+          v = (v & ((0x7f << 21) | (0x7f << 14) | (0x7f << 7) | 0x7f)) | ((*cur++) << 28);
+        }
+      }
+    }
+  }
+  return v;
+}
+
+// an individual batch. processed by a warp.
+// batches should be in shared memory.
+template <typename level_t>
+struct rle_batch {
+  uint8_t const* run_start;  // start of the run we are part of
+  int run_offset;            // value offset of this batch from the start of the run
+  level_t* output;
+  int level_run;
+  int size;
+
+  __device__ inline void decode(uint8_t const* const end, int level_bits, int lane, int warp_id)
+  {
+    int output_pos = 0;
+    int remain     = size;
+
+    // for bitpacked/literal runs, total size is always a multiple of 8. so we need to take care if
+    // we are not starting/ending exactly on a run boundary
+    uint8_t const* cur;
+    if (level_run & 1) {
+      int const effective_offset = cudf::util::round_down_safe(run_offset, 8);
+      int const lead_values      = (run_offset - effective_offset);
+      output_pos -= lead_values;
+      remain += lead_values;
+      cur = run_start + ((effective_offset >> 3) * level_bits);
+    }
+
+    // if this is a repeated run, compute the repeated value
+    int _level_val;
```

Comment:
While we're on underscores, is this prefixed so it won't be confused with a member?

Reply:
No, this is just relic dna from debugging.  Should probably just be level_val.  I'll change.